File: agent/internal/noise/handshake.go
// agent/internal/noise/handshake.go
package noise

import (
	"crypto/ecdh"
	"errors"
	"fmt"
)

type role int

const (
	roleInitiator role = iota
	roleResponder
)

type phase int

const (
	phaseReady phase = iota
	phaseMsg1Sent
	phaseMsg1Received
	phaseMsg2Sent
	phaseMsg2Received
	phaseDone
)

// NoiseXX implements the Noise_XX handshake state machine.
type NoiseXX struct {
	role  role
	phase phase
	h     [32]byte
	ck    [32]byte
	k     [32]byte // zero until first MixKey
	hasK  bool

	// Ephemeral keys (generated during handshake)
	ePriv *ecdh.PrivateKey
	ePub  []byte // 65 bytes

	// Static keys
	sPriv *ecdh.PrivateKey
	sPub  []byte // 65 bytes

	// Received remote keys
	rEPub []byte // remote ephemeral pub (65 bytes)
	rSPub []byte // remote static pub (65 bytes)
}

// NewInitiator creates a NoiseXX handshake as the initiator (browser).
// Generates a fresh per-session static key pair.
func NewInitiator() (*NoiseXX, error) {
	staticPriv, err := generateKeypair()
	if err != nil {
		return nil, fmt.Errorf("noise: generate initiator static: %w", err)
	}
	h, ck := initialize()
	h = mixHash(h, nil) // MixHash(prologue) where prologue = b""
	return &NoiseXX{
		role:  roleInitiator,
		phase: phaseReady,
		h:     h,
		ck:    ck,
		sPriv: staticPriv,
		sPub:  staticPriv.PublicKey().Bytes(),
	}, nil
}

// NewResponder creates a NoiseXX handshake as the responder (agent).
// Uses the agent's long-lived static key pair.
func NewResponder(staticPriv *ecdh.PrivateKey) (*NoiseXX, error) {
	h, ck := initialize()
	h = mixHash(h, nil) // MixHash(prologue) where prologue = b""
	return &NoiseXX{
		role:  roleResponder,
		phase: phaseReady,
		h:     h,
		ck:    ck,
		sPriv: staticPriv,
		sPub:  staticPriv.PublicKey().Bytes(),
	}, nil
}

// RemoteStaticPub returns the remote party's static public key (65 bytes).
// Only valid after ReadMessage2 (initiator) or ReadMessage3 (responder).
func (n *NoiseXX) RemoteStaticPub() []byte {
	return n.rSPub
}

// WriteMessage1 -- initiator sends: e (65 bytes)
func (n *NoiseXX) WriteMessage1() ([]byte, error) {
	if n.role != roleInitiator {
		return nil, errors.New("noise: WriteMessage1 called on responder")
	}
	if n.phase != phaseReady {
		return nil, fmt.Errorf("noise: WriteMessage1 invalid in phase %d", n.phase)
	}
	var err error
	n.ePriv, err = generateKeypair()
	if err != nil {
		return nil, fmt.Errorf("noise: generate ephemeral: %w", err)
	}
	n.ePub = n.ePriv.PublicKey().Bytes()
	n.h = mixHash(n.h, n.ePub) // token: e
	n.phase = phaseMsg1Sent
	return n.ePub, nil
}

// ReadMessage1 -- responder receives: e
func (n *NoiseXX) ReadMessage1(msg []byte) error {
	if n.role != roleResponder {
		return errors.New("noise: ReadMessage1 called on initiator")
	}
	if n.phase != phaseReady {
		return fmt.Errorf("noise: ReadMessage1 invalid in phase %d", n.phase)
	}
	if len(msg) != 65 {
		return fmt.Errorf("noise: ReadMessage1: expected 65 bytes, got %d", len(msg))
	}
	n.rEPub = msg
	n.h = mixHash(n.h, n.rEPub) // token: e
	n.phase = phaseMsg1Received
	return nil
}

// WriteMessage2 -- responder sends: e, ee, s, es (162 bytes)
func (n *NoiseXX) WriteMessage2() ([]byte, error) {
	if n.role != roleResponder {
		return nil, errors.New("noise: WriteMessage2 called on initiator")
	}
	if n.phase != phaseMsg1Received {
		return nil, fmt.Errorf("noise: WriteMessage2 invalid in phase %d", n.phase)
	}
	var err error
	n.ePriv, err = generateKeypair()
	if err != nil {
		return nil, fmt.Errorf("noise: generate ephemeral: %w", err)
	}
	n.ePub = n.ePriv.PublicKey().Bytes()

	// token: e
	n.h = mixHash(n.h, n.ePub)
	buf := make([]byte, 0, 162)
	buf = append(buf, n.ePub...) // 65 bytes

	// token: ee -- DH(re, ie)
	rEPubKey, err := importPublicKey(n.rEPub)
	if err != nil {
		return nil, fmt.Errorf("noise: import remote ephemeral: %w", err)
	}
	ee, err := dhP256(n.ePriv, rEPubKey)
	if err != nil {
		return nil, fmt.Errorf("noise: DH ee: %w", err)
	}
	n.mixKey(ee)

	// token: s -- EncryptAndHash(rs_pub)
	encS, err := n.encryptAndHash(n.sPub)
	if err != nil {
		return nil, fmt.Errorf("noise: EncryptAndHash s: %w", err)
	}
	buf = append(buf, encS...) // 81 bytes

	// token: es -- DH(rs, ie)
	rEPubKey2, err := importPublicKey(n.rEPub)
	if err != nil {
		return nil, err
	}
	es, err := dhP256(n.sPriv, rEPubKey2)
	if err != nil {
		return nil, fmt.Errorf("noise: DH es: %w", err)
	}
	n.mixKey(es)

	// empty payload
	tag, err := n.encryptAndHash([]byte{})
	if err != nil {
		return nil, err
	}
	buf = append(buf, tag...) // 16 bytes
	n.phase = phaseMsg2Sent
	return buf, nil
}

// ReadMessage2 -- initiator receives: e, ee, s, es (162 bytes)
func (n *NoiseXX) ReadMessage2(msg []byte) error {
	if n.role != roleInitiator {
		return errors.New("noise: ReadMessage2 called on responder")
	}
	if n.phase != phaseMsg1Sent {
		return fmt.Errorf("noise: ReadMessage2 invalid in phase %d", n.phase)
	}
	if len(msg) != 162 {
		return fmt.Errorf("noise: ReadMessage2: expected 162 bytes, got %d", len(msg))
	}

	// token: e
	n.rEPub = msg[:65]
	n.h = mixHash(n.h, n.rEPub)

	// token: ee -- DH(ie, re)
	rEPubKey, err := importPublicKey(n.rEPub)
	if err != nil {
		return fmt.Errorf("noise: import remote ephemeral: %w", err)
	}
	ee, err := dhP256(n.ePriv, rEPubKey)
	if err != nil {
		return fmt.Errorf("noise: DH ee: %w", err)
	}
	n.mixKey(ee)

	// token: s -- DecryptAndHash(encrypted_rs_pub)
	rsPub, err := n.decryptAndHash(msg[65:146])
	if err != nil {
		return fmt.Errorf("noise: DecryptAndHash s: %w", err)
	}
	n.rSPub = rsPub

	// token: es -- DH(ie, rs)
	rSPubKey, err := importPublicKey(rsPub)
	if err != nil {
		return fmt.Errorf("noise: import remote static: %w", err)
	}
	es, err := dhP256(n.ePriv, rSPubKey)
	if err != nil {
		return fmt.Errorf("noise: DH es: %w", err)
	}
	n.mixKey(es)

	// verify empty payload tag
	_, err = n.decryptAndHash(msg[146:162])
	if err == nil {
		n.phase = phaseMsg2Received
	}
	return err
}

// WriteMessage3 -- initiator sends: s, se (97 bytes)
func (n *NoiseXX) WriteMessage3() ([]byte, error) {
	if n.role != roleInitiator {
		return nil, errors.New("noise: WriteMessage3 called on responder")
	}
	if n.phase != phaseMsg2Received {
		return nil, fmt.Errorf("noise: WriteMessage3 invalid in phase %d", n.phase)
	}
	buf := make([]byte, 0, 97)

	// token: s -- EncryptAndHash(is_pub)
	encS, err := n.encryptAndHash(n.sPub)
	if err != nil {
		return nil, fmt.Errorf("noise: EncryptAndHash s: %w", err)
	}
	buf = append(buf, encS...) // 81 bytes

	// token: se -- DH(is, re)
	rEPubKey, err := importPublicKey(n.rEPub)
	if err != nil {
		return nil, err
	}
	se, err := dhP256(n.sPriv, rEPubKey)
	if err != nil {
		return nil, fmt.Errorf("noise: DH se: %w", err)
	}
	n.mixKey(se)

	// empty payload
	tag, err := n.encryptAndHash([]byte{})
	if err != nil {
		return nil, err
	}
	buf = append(buf, tag...) // 16 bytes
	n.phase = phaseDone
	return buf, nil
}

// ReadMessage3 -- responder receives: s, se (97 bytes)
func (n *NoiseXX) ReadMessage3(msg []byte) error {
	if n.role != roleResponder {
		return errors.New("noise: ReadMessage3 called on initiator")
	}
	if n.phase != phaseMsg2Sent {
		return fmt.Errorf("noise: ReadMessage3 invalid in phase %d", n.phase)
	}
	if len(msg) != 97 {
		return fmt.Errorf("noise: ReadMessage3: expected 97 bytes, got %d", len(msg))
	}

	// token: s -- DecryptAndHash(encrypted_is_pub)
	isPub, err := n.decryptAndHash(msg[:81])
	if err != nil {
		return fmt.Errorf("noise: DecryptAndHash s: %w", err)
	}
	n.rSPub = isPub

	// token: se -- DH(re, is)
	iSPubKey, err := importPublicKey(isPub)
	if err != nil {
		return fmt.Errorf("noise: import initiator static: %w", err)
	}
	se, err := dhP256(n.ePriv, iSPubKey)
	if err != nil {
		return fmt.Errorf("noise: DH se: %w", err)
	}
	n.mixKey(se)

	// verify empty payload tag
	_, err = n.decryptAndHash(msg[81:97])
	if err == nil {
		n.phase = phaseDone
	}
	return err
}

// Split derives post-handshake cipher states.
// Returns (send, recv) from the caller's perspective.
func (n *NoiseXX) Split() (send, recv *CipherState) {
	if n.phase != phaseDone {
		panic("noise: Split called before handshake completion")
	}
	k1, k2 := splitKeys(n.ck)
	cs1, cs2 := newCipherState(k1), newCipherState(k2)
	if n.role == roleInitiator {
		return cs1, cs2 // initiator sends on k1, receives on k2
	}
	return cs2, cs1 // responder sends on k2, receives on k1
}

// mixKey updates ck and k from a DH output (Noise spec §5.2 MixKey).
func (n *NoiseXX) mixKey(ikm []byte) {
	n.ck, n.k = hkdf2(n.ck, ikm)
	n.hasK = true
}

// encryptAndHash: ciphertext = EncryptWithAd(h, pt); MixHash(ciphertext)
func (n *NoiseXX) encryptAndHash(plaintext []byte) ([]byte, error) {
	if !n.hasK {
		return nil, errors.New("noise: cipher key not initialized")
	}
	cs := &CipherState{key: n.k}
	ct, err := cs.encryptWithAd(n.h[:], plaintext)
	if err != nil {
		return nil, err
	}
	n.h = mixHash(n.h, ct)
	return ct, nil
}

// decryptAndHash: plaintext = DecryptWithAd(h, ct); MixHash(ciphertext)
func (n *NoiseXX) decryptAndHash(ciphertext []byte) ([]byte, error) {
	if !n.hasK {
		return nil, errors.New("noise: cipher key not initialized")
	}
	cs := &CipherState{key: n.k}
	pt, err := cs.decryptWithAd(n.h[:], ciphertext)
	if err != nil {
		return nil, err
	}
	n.h = mixHash(n.h, ciphertext)
	return pt, nil
}

File: control/web/noise-p256/handshake.js
// signaling-server/web/noise-p256/handshake.js
import { initialize, mixHash, hkdf2 } from './transcript.js'
import { CipherState, splitKeys } from './cipher.js'
import { generateKeypair, dh, importPublicKey, exportPublicKey } from './keys.js'

export class NoiseXX {
  #role  // 'initiator' | 'responder'
  #phase // 'ready' | 'msg1-sent' | 'msg1-received' | 'msg2-sent' | 'msg2-received' | 'done'
  #h
  #ck
  #k = new Uint8Array(32)  // zero until first mixKey
  #hasK = false

  #ePriv = null   // ephemeral private key (CryptoKey)
  #ePub = null    // ephemeral public key bytes (65 bytes)
  #sPriv = null   // static private key (CryptoKey)
  #sPub = null    // static public key bytes (65 bytes)
  #rEPub = null   // remote ephemeral pub bytes
  #rSPub = null   // remote static pub bytes

  constructor(role, sPriv, sPub) {
    this.#role = role
    this.#phase = 'ready'
    const { h, ck } = initialize()
    this.#h = h
    this.#ck = ck
    this.#sPriv = sPriv
    this.#sPub = sPub
  }

  static async createInitiator(opts = {}) {
    const kp = opts.staticKeypair ?? await generateKeypair()
    const noise = new NoiseXX('initiator', kp.privateKey, kp.publicKeyBytes)
    noise._testEphemeralKeypair = opts.ephemeralKeypair ?? null
    noise.#h = await mixHash(noise.#h, new Uint8Array(0)) // MixHash(prologue)
    return noise
  }

  static async createResponder(staticPrivKey, staticPubKeyBytes, opts = {}) {
    const noise = new NoiseXX('responder', staticPrivKey, staticPubKeyBytes)
    noise._testEphemeralKeypair = opts.ephemeralKeypair ?? null
    noise.#h = await mixHash(noise.#h, new Uint8Array(0)) // MixHash(prologue)
    return noise
  }

  get remoteStaticPub() { return this.#rSPub }

  // writeMessage1 — initiator: e (65 bytes)
  async writeMessage1() {
    this._assertRole('initiator', 'writeMessage1')
    this._assertPhase('ready', 'writeMessage1')
    const kp = this._testEphemeralKeypair ?? await generateKeypair()
    this._testEphemeralKeypair = null
    this.#ePriv = kp.privateKey
    this.#ePub = kp.publicKeyBytes
    this.#h = await mixHash(this.#h, this.#ePub)
    this.#phase = 'msg1-sent'
    return this.#ePub.slice()
  }

  // readMessage1 — responder: e
  async readMessage1(msg) {
    this._assertRole('responder', 'readMessage1')
    this._assertPhase('ready', 'readMessage1')
    this._assertLen(msg, 65, 'readMessage1')
    this.#rEPub = msg.slice(0, 65)
    this.#h = await mixHash(this.#h, this.#rEPub)
    this.#phase = 'msg1-received'
  }

  // writeMessage2 — responder: e, ee, s, es (162 bytes)
  async writeMessage2() {
    this._assertRole('responder', 'writeMessage2')
    this._assertPhase('msg1-received', 'writeMessage2')
    const kp = this._testEphemeralKeypair ?? await generateKeypair()
    this._testEphemeralKeypair = null
    this.#ePriv = kp.privateKey
    this.#ePub = kp.publicKeyBytes

    // token: e
    this.#h = await mixHash(this.#h, this.#ePub)
    const buf = [this.#ePub]

    // token: ee
    const ee = await dh(this.#ePriv, this.#rEPub)
    await this.#mixKey(ee)

    // token: s — EncryptAndHash(rs_pub)
    const encS = await this.#encryptAndHash(this.#sPub)
    buf.push(encS)

    // token: es — DH(rs, ie)
    const es = await dh(this.#sPriv, this.#rEPub)
    await this.#mixKey(es)

    // empty payload
    const tag = await this.#encryptAndHash(new Uint8Array(0))
    buf.push(tag)
    this.#phase = 'msg2-sent'
    return concat(...buf)
  }

  // readMessage2 — initiator: e, ee, s, es (162 bytes)
  async readMessage2(msg) {
    this._assertRole('initiator', 'readMessage2')
    this._assertPhase('msg1-sent', 'readMessage2')
    this._assertLen(msg, 162, 'readMessage2')

    // token: e
    this.#rEPub = msg.slice(0, 65)
    this.#h = await mixHash(this.#h, this.#rEPub)

    // token: ee
    const ee = await dh(this.#ePriv, this.#rEPub)
    await this.#mixKey(ee)

    // token: s — DecryptAndHash(encrypted_rs_pub)
    this.#rSPub = await this.#decryptAndHash(msg.slice(65, 146))

    // token: es — DH(ie, rs)
    const es = await dh(this.#ePriv, this.#rSPub)
    await this.#mixKey(es)

    // verify empty payload
    await this.#decryptAndHash(msg.slice(146, 162))
    this.#phase = 'msg2-received'
  }

  // writeMessage3 — initiator: s, se (97 bytes)
  async writeMessage3() {
    this._assertRole('initiator', 'writeMessage3')
    this._assertPhase('msg2-received', 'writeMessage3')
    const buf = []

    // token: s — EncryptAndHash(is_pub)
    const encS = await this.#encryptAndHash(this.#sPub)
    buf.push(encS)

    // token: se — DH(is, re)
    const se = await dh(this.#sPriv, this.#rEPub)
    await this.#mixKey(se)

    // empty payload
    const tag = await this.#encryptAndHash(new Uint8Array(0))
    buf.push(tag)
    this.#phase = 'done'
    return concat(...buf)
  }

  // readMessage3 — responder: s, se (97 bytes)
  async readMessage3(msg) {
    this._assertRole('responder', 'readMessage3')
    this._assertPhase('msg2-sent', 'readMessage3')
    this._assertLen(msg, 97, 'readMessage3')

    // token: s — DecryptAndHash(encrypted_is_pub)
    this.#rSPub = await this.#decryptAndHash(msg.slice(0, 81))

    // token: se — DH(re, is)
    const se = await dh(this.#ePriv, this.#rSPub)
    await this.#mixKey(se)

    // verify empty payload
    await this.#decryptAndHash(msg.slice(81, 97))
    this.#phase = 'done'
  }

  // split returns [send, recv] from the caller's perspective.
  async split() {
    this._assertPhase('done', 'split')
    const [k1, k2] = await splitKeys(this.#ck)
    const cs1 = new CipherState(k1)
    const cs2 = new CipherState(k2)
    return this.#role === 'initiator' ? [cs1, cs2] : [cs2, cs1]
  }

  async #mixKey(ikm) {
    const [newCK, newK] = await hkdf2(this.#ck, ikm)
    this.#ck = newCK
    this.#k = newK
    this.#hasK = true
  }

  async #encryptAndHash(plaintext) {
    if (!this.#hasK) throw new Error('noise: cipher key not initialized')
    const cs = new CipherState(this.#k)
    const ct = await cs.encryptWithAd(this.#h, plaintext)
    this.#h = await mixHash(this.#h, ct)
    return ct
  }

  async #decryptAndHash(ciphertext) {
    if (!this.#hasK) throw new Error('noise: cipher key not initialized')
    const cs = new CipherState(this.#k)
    const pt = await cs.decryptWithAd(this.#h, ciphertext)
    this.#h = await mixHash(this.#h, ciphertext)
    return pt
  }

  _assertRole(expected, method) {
    if (this.#role !== expected) throw new Error(`noise: ${method} called on ${this.#role}`)
  }

  _assertPhase(expected, method) {
    if (this.#phase !== expected) throw new Error(`noise: ${method} invalid in phase ${this.#phase}`)
  }

  _assertLen(buf, expected, method) {
    if (buf.length !== expected) throw new Error(`noise: ${method}: expected ${expected} bytes, got ${buf.length}`)
  }
}

function concat(...arrays) {
  const total = arrays.reduce((n, a) => n + a.length, 0)
  const out = new Uint8Array(total)
  let offset = 0
  for (const arr of arrays) { out.set(arr, offset); offset += arr.length }
  return out
}